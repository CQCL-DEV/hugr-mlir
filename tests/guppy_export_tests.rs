
use rstest::{rstest,fixture};

use hugr_mlir::Result;

#[fixture]
pub fn test_context() -> melior::Context {
    let dr = melior::dialect::DialectRegistry::new();
    let hugr_dh = hugr_mlir::mlir::hugr::get_hugr_dialect_handle();
    hugr_dh.insert_dialect(&dr);
    let ctx = melior::Context::new();
    ctx.append_dialect_registry(&dr);
    ctx.load_all_available_dialects();
    ctx
}

#[rstest]
fn test_guppy_exports(test_context: melior::Context) -> Result<()> {
    insta::glob!("guppy-exports", "*.json", |path| {
        use std::fs;
        let src = fs::read(path).unwrap();
        let ul = melior::ir::Location::new(&test_context, path.to_str().unwrap(), 0, 0);
        let m = hugr_mlir::translate::translate_hugr_to_mlir(&src, ul, |bytes| serde_json::from_slice::<hugr::Hugr>(bytes)).unwrap();
        let o = m.as_operation();
        println!("{o}");
        assert!(m.as_operation().verify());
        insta::assert_display_snapshot!(m.as_operation());
    });
    Ok(())
}
