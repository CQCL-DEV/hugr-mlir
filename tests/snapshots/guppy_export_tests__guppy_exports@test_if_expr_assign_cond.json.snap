---
source: tests/guppy_export_tests.rs
expression: m.as_operation()
input_file: tests/guppy-exports/test_if_expr_assign_cond.json
---
module {
  hugr.func @foo[](%arg0: !hugr.sum<tuple<>, tuple<>>) -> !hugr.sum<tuple<>, tuple<>> {
    %0 = cfg %arg0 : (!hugr.sum<tuple<>, tuple<>>) -> !hugr.sum<tuple<>, tuple<>> {
    ^bb0(%arg1: !hugr.sum<tuple<>, tuple<>>):
      %1 = make_tuple(%arg1 : !hugr.sum<tuple<>, tuple<>>)
      %2 = make_tuple()
      %3 = conditional(%arg1, %1, %2 : !hugr.sum<tuple<>, tuple<>>, tuple<!hugr.sum<tuple<>, tuple<>>>, tuple<>) -> !hugr.sum<tuple<!hugr.sum<tuple<>, tuple<>>>, tuple<>> {
      ^bb0(%arg2: tuple<!hugr.sum<tuple<>, tuple<>>>, %arg3: tuple<>):
        %14 = tag 0 %arg2 : tuple<!hugr.sum<tuple<>, tuple<>>> -> <tuple<!hugr.sum<tuple<>, tuple<>>>, tuple<>>
        output %14 : !hugr.sum<tuple<!hugr.sum<tuple<>, tuple<>>>, tuple<>>
      }, {
      ^bb0(%arg2: tuple<!hugr.sum<tuple<>, tuple<>>>, %arg3: tuple<>):
        %14 = tag 1 %arg3 : tuple<> -> <tuple<!hugr.sum<tuple<>, tuple<>>>, tuple<>>
        output %14 : !hugr.sum<tuple<!hugr.sum<tuple<>, tuple<>>>, tuple<>>
      }
      switch %3 : !hugr.sum<tuple<!hugr.sum<tuple<>, tuple<>>>, tuple<>> ^bb2, ^bb3
    ^bb1(%4: !hugr.sum<tuple<>, tuple<>>):  // pred: ^bb4
      output %4 : !hugr.sum<tuple<>, tuple<>>
    ^bb2(%5: !hugr.sum<tuple<>, tuple<>>):  // pred: ^bb0
      %6 = make_tuple()
      %7 = tag 0 %6 : tuple<> -> <tuple<>>
      switch %7, %5 : !hugr.sum<tuple<>>, !hugr.sum<tuple<>, tuple<>> ^bb4
    ^bb3:  // pred: ^bb0
      const @"const_Node { index: NodeIndex(30) }" : !hugr.sum<tuple<>, tuple<>> = #hugr.sum<0, []> : !hugr.sum<tuple<>, tuple<>>
      %8 = load_constant @"const_Node { index: NodeIndex(30) }" : !hugr.sum<tuple<>, tuple<>>
      %9 = make_tuple()
      %10 = tag 0 %9 : tuple<> -> <tuple<>>
      switch %10, %8 : !hugr.sum<tuple<>>, !hugr.sum<tuple<>, tuple<>> ^bb4
    ^bb4(%11: !hugr.sum<tuple<>, tuple<>>):  // 2 preds: ^bb2, ^bb3
      %12 = make_tuple()
      %13 = tag 0 %12 : tuple<> -> <tuple<>>
      switch %13, %11 : !hugr.sum<tuple<>>, !hugr.sum<tuple<>, tuple<>> ^bb1
    }
    output %0 : !hugr.sum<tuple<>, tuple<>>
  }
}

